editorial: Tidy up uses of "in parallel".

rakuco opened the pull request:
Steps that run "in parallel" need to take several extra restrictions into
account: we do not have access to documents or global objects, cannot create
IDL interfaces or manipulate promises, for example, all of which we had been
doing, along with needlessly running nested "in parallel" steps.

This change attempts to rectify the situation by only running things in
parallel when absolutely necessary (i.e. when reaching out to the OS), and
queuing global tasks (emphasis on "global" given whatwg/html#4980 and
whatwg/html#5653) from there when we need to manipulate promises or create
objects.

"Obtain permission" algorithm:
* Stop running in parallel. Callers should be responsible for choosing
  whether it should be run in parallel or not.

`WakeLock.request()`:
* Separate returning a promise and running steps in parallel. This style is
  more usual.
* Refer to the "relevant settings object" rather than the "current settings
  object", as we are inside a method of an IDL interface and can rely on it
  being defined. I do not think this is a user-visible change, and it looks
  cleaner.
* Queue a global task to reject `|promise|` when the permission request run
  in parallel is denied.
* Manipulate the `[[ActiveLocks]]` internal slot, check page visibility,
  invoke "acquire a wake lock", create a new WakeLockSentinel and resolve
  the returned promise in a queued global task.

`WakeLockSentinel.release()`:
* Do not run the "release a wake lock" algorithm in parallel (see the
  changes to the algorithm itself below).
* Just return a resolved promise once the rest of the steps run. The
  returned promise does not have much use, but has been kept to avoid
  breaking API compatibility.
* One user-visible consequence is that the "release" event is fired
  synchronously and before the function returns.

"Acquire wake lock" algorithm:
* Stop running everything in parallel.
* Move all `[[ActiveLocks]]` manipulation to `WakeLock.request()` itself,
  and make the algorithm just ask the platform for a wake lock.

"Release wake lock" algorithm:
* Stop running everything in parallel. The only steps that still need to run
  in parallel are the ones asking the platform to release the wake lock.
* Consequently, stop queueing a task to change `[[Released]]` and fire the
  "release" event, and do it directly in the algorithm instead.

Big thanks to Anne van Kesteren, Domenic Denicola and Marcos Cáceres for
their input.

Fixes #293.

---

### Open items:
* [x] How does one check if `|document|` is hidden in `WakeLock.request()`'s parallel steps? Or, more generally speaking, how can we avoid reaching the steps where we ask the platform for a lock if the page was hidden, for example?
* [x] `WakeLock.request()`: does it make sense to have to separate "in parallel" steps to first acquire a platform lock and then manipulate `[[ActiveLocks]]`?
* [x] `WakeLock.request()`: does manipulating `[[ActiveLocks]]` from a queued task leave the algorithm open to race conditions?
* [x] `WakeLockSentinel.release()`: the [Web Platform Design Principles](https://w3ctag.github.io/design-principles/#promises-and-events) say "events should fire before Promises resolve". Do the changes make sense here? We are changing `[[Released]]` and firing the "released" event "synchronously" in the event loop before even creating the resolved promise we are returning. This feels a bit odd, but on the other hand I'm not sure how to avoid running the "release wake lock" algorithm in parallel if we create the promise first.
  * This one of the things that led to #293: if we queue a task to fire the event, it might not be delivered at all. Is that OK? Should `[[Released]]` be changed synchronously though, so that e.g. when we are navigating to another page, should `.released` is true in the "visibilitychange" event handler even though the "release" event has not been dispatched?

---

The following tasks have been completed:

 * [x] Modified Web platform tests (https://github.com/web-platform-tests/wpt/pull/28148)

Implementation commitment:

 * [x] Chromium (https://bugs.chromium.org/p/chromium/issues/detail?id=1190110) 
 * WebKit - Not participating in this working group.


<!--
    This comment and the below content is programatically generated.
    You may add a comma-separated list of anchors you'd like a
    direct link to below (e.g. #idl-serializers, #idl-sequence):

    Don't remove this comment or modify anything below this line.
    If you don't want a preview generated for this pull request,
    just replace the whole of this comment's content by "no preview"
    and remove what's below.
-->
***
<a href="https://pr-preview.s3.amazonaws.com/rakuco/wake-lock/pull/299.html" title="Last updated on Mar 22, 2021, 11:43 AM UTC (7ac57c1)">Preview</a> | <a href="https://pr-preview.s3.amazonaws.com/w3c/screen-wake-lock/299/9fb419f...rakuco:7ac57c1.html" title="Last updated on Mar 22, 2021, 11:43 AM UTC (7ac57c1)">Diff</a>

Commit:
editorial: Tidy up uses of "in parallel".

Steps that run "in parallel" need to take several extra restrictions into
account: we do not have access to documents or global objects, cannot create
IDL interfaces or manipulate promises, for example, all of which we had been
doing, along with needlessly running nested "in parallel" steps.

This change attempts to rectify the situation by only running things in
parallel when absolutely necessary (i.e. when reaching out to the OS), and
queuing global tasks (emphasis on "global" given whatwg/html#4980 and
whatwg/html#5653) from there when we need to manipulate promises or create
objects.

"Obtain permission" algorithm:
* Stop running in parallel. Callers should be responsible for choosing
  whether it should be run in parallel or not.

`WakeLock.request()`:
* Separate returning a promise and running steps in parallel. This style is
  more usual.
* Refer to the "relevant settings object" rather than the "current settings
  object", as we are inside a method of an IDL interface and can rely on it
  being defined. I do not think this is a user-visible change, and it looks
  cleaner.
* Queue a global task to reject `|promise|` when the permission request run
  in parallel is denied.
* Manipulate the `[[ActiveLocks]]` internal slot, check page visibility,
  invoke "acquire a wake lock", create a new WakeLockSentinel and resolve
  the returned promise in a queued global task.

`WakeLockSentinel.release()`:
* Do not run the "release a wake lock" algorithm in parallel (see the
  changes to the algorithm itself below).
* Just return a resolved promise once the rest of the steps run. The
  returned promise does not have much use, but has been kept to avoid
  breaking API compatibility.
* One user-visible consequence is that the "release" event is fired
  synchronously and before the function returns.

"Acquire wake lock" algorithm:
* Stop running everything in parallel.
* Move all `[[ActiveLocks]]` manipulation to `WakeLock.request()` itself,
  and make the algorithm just ask the platform for a wake lock.

"Release wake lock" algorithm:
* Stop running everything in parallel. The only steps that still need to run
  in parallel are the ones asking the platform to release the wake lock.
* Consequently, stop queueing a task to change `[[Released]]` and fire the
  "release" event, and do it directly in the algorithm instead.

Big thanks to Anne van Kesteren, Domenic Denicola and Marcos Cáceres for
their input.

Fixes #293.

rakuco commented:
@annevk @domenic this is a first attempt to tackle #293 after the discussion we've had there. I'd appreciate it if you could see if this makes sense and look at the "open items" in the PR message with things that are still unclear to me.

kenchris commented on index.html:
```diff
@@ -508,19 +516,12 @@ <h3>
-              </li>
-            </ol>
+          <li>Otherwise, run <a>release a wake lock</a> with
+          |lock:WakeLockSentinel| set to this object and |type:WakeLockType|
```
not sure I understand "set to this  object"

rakuco replied:
Note that "this object" is the language we've already been using here and in other steps. If just saying `this` is more idiomatic, I can change all occurrences in a separate PR.

kenchris replied:
Só why not use backtick ` around 'this'

rakuco replied:
The short answer is "because nobody had ever brought it up before" :-) I'm just following the language we had already been using -- this excerpt was reindented but otherwise left completely unmodified. I can change that, but I'd rather do it in a separate PR because we say that in other places too.

rakuco replied:
Marking this as resolved, this change was landed in #304

kenchris commented on index.html:
```diff
@@ -508,19 +516,12 @@ <h3>
-            </ol>
+          <li>Otherwise, run <a>release a wake lock</a> with
+          |lock:WakeLockSentinel| set to this object and |type:WakeLockType|
+          set to the value of this object's {{WakeLockSentinel/type}}
```
what object? "`this`"?

rakuco replied:
Yes. The `WakeLockSentinel` object whose `release()` method is being invoked.

reillyeon commented:
While you are here fixing the "in parallel" sections this specification should be updated to define a task source onto which the tasks returning to main thread are queued.

rakuco commented:
> While you are here fixing the "in parallel" sections this specification should be updated to define a task source onto which the tasks returning to main thread are queued.

Do you think it's better to come up with a new task source, or would it make sense to use the [user interaction task source](https://html.spec.whatwg.org/multipage/webappapis.html#user-interaction-task-source)?

rakuco commented:
> Do you think it's better to come up with a new task source, or would it make sense to use the [user interaction task source](https://html.spec.whatwg.org/multipage/webappapis.html#user-interaction-task-source)?

Per w3c/presentation-api#360 and w3ctag/design-principles#38 it looks like it makes more sense to define another task source.

rakuco commented:
@annevk @domenic friendly ping, it'd be great if you could take a look at the proposed changes here and/or the questions in the PR message which I couldn't answer yet.

annevk commented:
Some thoughts:

* obtain permission shouldn't run in parallel, it should be invoked from in parallel, but it doesn't make sense for it itself to then also go in parallel again (in parallel steps cannot return things)
* A slightly clearer style is to say "run these steps in parallel" and then have a subsequent step return the promise.
* You really want to use "queue a global task" and always pass it the global and task source to use explicitly. That's the way forward.

In parallel usage otherwise seems reasonable from a quick skim. (If you're still not too sure about it, HTML has a bunch of guidance language for it, it's basically akin to a thread/process. Also happy to chat at https://whatwg.org/irc or equivalent.)

Questions:

* I'm not sure about hidden documents. There are some issues about this against HTML as well. (Checking the `hidden` attribute is definitely wrong though. You don't want to invoke IDL members in your algorithms as a general rule.)
* I'm not sure I understand this.
* Yeah, it seems "acquire a wake lock" could be run twice if you invoke the method twice.
* The order of event/promise in `release()` seems fine. It might be clearer to run "release a wake lock" when the slot is false and then have a single return step at the end. 
   * If you don't want `release()` to fire an event directly (though not sure if it matters much), you could make it queue a task to run "release a wake lock" and resolve the promise with undefined. That way other callers of "release a wake lock" would still dispatch the event in whatever task they find themselves in.

rakuco commented:
Thanks for the review, @annevk. I've pushed some new commits addressing some of the points you've brought up.

> Some thoughts:
> 
> * obtain permission shouldn't run in parallel, it should be invoked from in parallel, but it doesn't make sense for it itself to then also go in parallel again (in parallel steps cannot return things)

I was hoping to get rid of the algorithm altogether in #298, but in the meantime I've dropped the "in parallel" bit in this PR.

> * A slightly clearer style is to say "run these steps in parallel" and then have a subsequent step return the promise.

Done.

> * You really want to use "queue a global task" and always pass it the global and task source to use explicitly. That's the way forward.

Is there any particular reason for using "global" here? I recently defined a task source in #301 and was using the regular "queue a task" language with it.

> Questions:
> 
> * I'm not sure about hidden documents. There are some issues about this against HTML as well. (Checking the `hidden` attribute is definitely wrong though. You don't want to invoke IDL members in your algorithms as a general rule.)

Hmm, I wonder if the algorithms in "handling document loss of {full activity, visibility}" could set some flag that could be checked in the parallel steps of `WakeLock.request()`. Some sort of "_abort when_ some-flag is set // _if aborted_ reject promise".

> * Yeah, it seems "acquire a wake lock" could be run twice if you invoke the method twice.

Would it be OK (from an idiomatic perspective) to remove the check altogether and sort of under-specify the behavior by leaving it up to each implementation to determine whether to ask the platform for a lock?

> * The order of event/promise in `release()` seems fine. It might be clearer to run "release a wake lock" when the slot is false and then have a single return step at the end.

Thanks, done.
  
>   * If you don't want `release()` to fire an event directly (though not sure if it matters much), you could make it queue a task to run "release a wake lock" and resolve the promise with undefined. That way other callers of "release a wake lock" would still dispatch the event in whatever task they find themselves in.

I'm trying to determine if there's some general guidance here. If it's basically "fire events away and only queue a task to fire them if necessary", then I'm fine with the changes being done here.

annevk commented:
> Is there any particular reason for using "global" here?

Yes, it helps ensure the task is queued on the correct event loop and all the relevant bits are set correctly. (Though @domenic has been suggesting that we have some kind of ambient global available for constructing objects we could also repurpose here perhaps...)

> Would it be OK (from an idiomatic perspective) to remove the check altogether and sort of under-specify the behavior by leaving it up to each implementation to determine whether to ask the platform for a lock?

Maybe, not sure. It seems easy enough to add some boolean that you can use to prevent this, no?

> If it's basically "fire events away and only queue a task to fire them if necessary", then I'm fine with the changes being done here.

It depends on the cause and the target, I think here it's fine. If it was in response to a DOM mutation I'd ask for a change.

rakuco commented:
@annevk I've pushed a few more commits. I think I've managed to solve everything, but I am not entirely sure doing things from a queued global task is enough to avoid race conditions and not being allowed to refer to global objects:
* Switched to "queue a global task"
* Switched to manipulating the `[[ActiveLocks]]` list only in the main event loop (from a queued global task in `WakeLock.request()` and without doing anything special in the "Release a wake lock" algorithm)
* I think I've solved the page visibility issue by checking it inside the queued global task in `WakeLock.request()` (I'm adjusting the wording more broadly in #306).

marcoscaceres commented on index.html:
```diff
@@ -361,48 +365,67 @@ <h3>
-              <li>Resolve |promise| with |lock|.
+              <li>
+                <a>Queue a global task</a> on the <a>screen wake lock task
+                source</a> given |global| to run these steps:
```
Not totally clear what "given |global|" means here.

rakuco replied:
It's defined in the beginning of the steps as "`this's` relevant object", is there a clearer way to refer to it?

kenchris replied:
domenic used wording like "given |document|'s [=relevant global object=] to"

rakuco replied:
Yeah, I noticed shortly after posting the comment above :-D I've pushed a new version incorporating that.

marcoscaceres commented on index.html:
```diff
@@ -506,22 +529,12 @@ <h3>
-              <li>Resolve |promise|.
-              </li>
-            </ol>
+          <li>If <a>this</a>' {{[[Released]]}} internal slot is `false`, then
```
```suggestion
          <li>If <a>this</a>' {{[[Released]]}} is `false`, then
```

rakuco replied:
Fixed.

marcoscaceres commented on index.html:
```diff
@@ -767,15 +768,10 @@ <h3>
               </li>
             </ol>
           </li>
+          <li>Set |lock|'s {{[[Released]]}} internal slot to `true`.
```
```suggestion
          <li>Set |lock|'s {{[[Released]]}} to `true`.
```

rakuco replied:
Done.

marcoscaceres commented on index.html:
```diff
@@ -861,8 +857,8 @@ <h2>
         Konstantinov, Matvey Larionov, Dominique Hazael-Massieux, Domenic
-        Denicola, Thomas Steiner, Raphael Kubo da Costa for their contributions
-        to this work.
+        Denicola, Thomas Steiner, Raphael Kubo da Costa, Anne van Kesteren for
```
No need to thank an editor ;) 

```suggestion
        Denicola, Thomas Steiner, Anne van Kesteren for
```

rakuco replied:
Done :-)

marcoscaceres commented on index.html:
```diff
@@ -767,15 +768,10 @@ <h3>
-                <a>Fire an event</a> named "`release`" at |lock|.
-              </li>
-            </ol>
+            <a>Fire an event</a> named "`release`" at |lock|.
```
Seems odd that this is not queued? or was it queued above and I missed it?

rakuco replied:
Heh, that's how this entire PR started: I asked in #293 whether it made sense to queue the task, and we uncovered a lot of issues with "in parallel" in the spec. I also asked above and it seemed like there was no need to queue this at all since we're not running in parallel.

domenic replied:
So not queuing this means that the following code should work:

```js
let fired = false;
lock.addEventListener("release", () => fired = true);
lock.release();
console.assert(fired === true);
```

Are you sure that's how things work? It seems possible, but I just wanted to check.

rakuco replied:
I _think_ so? I'm honestly fine either way, it's just no clear to me if there's a guideline for deciding whether to queue a task or not, and the fact that `WakeLockSentinel.release()` returns a promise also makes it more confusing:
* Should the snippet above work, or should one have to wait for the promise to resolve?
* Doesn't queuing a task mean it will be fired after the promise resolves?
* If the task is queued, it will likely not be run when the page is being unloaded (as part of the "handling loss of page visibility" section), while if we run these steps directly here they will. Is there a preference?

domenic replied:
I mean, can you write a test?

Really it seems like `release()` shouldn't return a promise at all, since it's not giving any meaningful information. I wonder if that would be a backward-compatible change.

reillyeon replied:
I agree that there is no value in `release()` returning a `Promise` but it would not be a backwards-compatible change since `lock.release().then()` would throw an error if it returned `undefined`.

domenic replied:
It might be worth adding a note to the spec saying that the returned promise is a historical artifact and only retained for backward compatibility.

rakuco replied:
One thing we can also do is make this part of the change adhere to ["Events should fire before Promises resolve"](https://w3ctag.github.io/design-principles/#promises-and-events) and only resolve the promise after firing the event. The existing version of the spec kind of attempts to do that, but since we queue a task while already in parallel to fire the event I'm not sure this is really guaranteed to happen.

In any case: is there a guideline for deciding whether to queue a task or not when firing an event inside the main event loop?

domenic replied:
You generally don't queue a task for synchronous things, like this one.

I don't think the guidance is applicable in this case since it's a synchronous thing that only returns a promise for historical reasons.

annevk replied:
If you fire the event and resolve the promise in the same task, the event will fire before the promise resolves, right? So if you do that in the task that invokes `release()` (i.e., `release()` doesn't queue a task itself and just fires the event and resolves the promise) I think you would be adhering to that, even if slightly weird and probably not to be repeated.

rakuco replied:
Yeah, in which case the current patch would be fine as-is. I've read #226 and #237 again to see if there was any explanation for making `release()` return a Promise, but couldn't find anything other than the suggested Web IDL snippets from #226. It looks weird right now, but I guess it also allows us to do more things in the future if necessary and use it to signal everything has finished running?

annevk replied:
Yeah, maybe.

marcoscaceres replied:
Thanks for the clarifications on this. Makes sense now. @rakuco, feel free to resolve this if you are happy.

marcoscaceres reviewed (commented):
Some drive by feedback...

kenchris commented on index.html:
```diff
@@ -337,8 +337,11 @@ <h3>
           <li>Let |document:Document| be the [=environment settings object /
-          responsible document=] of the <a>current settings object</a>.
+          responsible document=] of <a>this</a>' <a>relevant settings
+          object</a>.
```
Domenic just used "<li>Let |document:Document| be the [=current settings object=]'s [=associated Document=]."

rakuco replied:
I think it makes sense to clean this up in #307 rather than here, as it's part of the effort to stop mentioning "responsible document" altogether.

rakuco commented:
@annevk @domenic @kenchris @marcoscaceres I think this PR is finally ready. All uses of "in parallel" have been audited, I think `WakeLock.request()` is free of race conditions when manipulating `[[ActiveLocks]]` and we are not checking anything from the main event loop when in parallel. The questions that seems to remain is there should be a queued task to fire the "release" event and change `[[Released]]` to `true`, which is what motivated #293 in the first place. My understanding based on [this comment](https://github.com/w3c/screen-wake-lock/pull/299#issuecomment-777600796) above is that in this case there is no need to queue a task and we can do things directly instead.

There are further cleanups that @domenic is doing in #307 related to the use of "responsible document"; I think it makes sense to land them separately there.

domenic commented on index.html:
```diff
@@ -338,7 +338,8 @@ <h3>
         <ol class="algorithm">
           <li>Let |document:Document| be the [=environment settings object /
-          responsible document=] of the <a>current settings object</a>.
+          responsible document=] of <a>this</a>' <a>relevant settings
```
Nit: here and elsewhere, `this's` instead of `this'`.

rakuco replied:
Done in #309 (I'll rebase this and adjust the remaining occurrences after it lands).

domenic commented on index.html:
```diff
@@ -367,45 +368,64 @@ <h3>
           return [=a promise rejected with=] {{"NotAllowedError"}}
           {{DOMException}}.
           </li>
```
Nit: steps 1.1 and 1.2 should be steps 2 and 3, i.e. they should not be indented.

rakuco replied:
Done in #308.

domenic commented on index.html:
```diff
@@ -367,45 +368,64 @@ <h3>
+              <li>Let |state:PermissionState| be the result of invoking
+              <a>obtain permission</a> with "`screen-wake-lock`":
                 <ol>
                   <li data-tests="wakelock-request-denied.https.html">If
```
This step should not be nested.

rakuco replied:
Done in #308.

domenic commented on index.html:
```diff
@@ -367,45 +368,64 @@ <h3>
+                  given |document|'s <a>relevant global object</a> to reject
+                  |promise| with a {{"NotAllowedError"}} {{DOMException}}
+                  </li>
+                  <li>Abort these steps.
```
This aborts no matter what. You probably want to nest this underneath the "If...denied" as a sibling of "Queue a global task".

rakuco replied:
Oops, you're right. I tried to incorporate some of the changes from #307 but didn't pay attention to this one. Fixed in #308 and here.

domenic commented on index.html:
```diff
@@ -749,8 +748,7 @@ <h3>
           </li>
-          <li>If the internal slot {{[[ActiveLocks]]}} of all the <a>platform
-          wake lock</a>'s <a>state record</a>s are all empty, then run the
+          <li>If |record|.{{[[ActiveLocks]]}} is [=list/empty=], then run the
```
```suggestion
          <li>If |record|.{{[[ActiveLocks]]}} [=list/is empty=], then run the
```

rakuco replied:
Fixed, thanks.

domenic commented on index.html:
```diff
@@ -367,45 +368,64 @@ <h3>
+                  return `hidden`, reject |promise| with a
+                  {{"NotAllowedError"}} {{DOMException}}.
+                  </li>
+                  <li>If |record|.{{[[ActiveLocks]]}} is [=list/empty=], then
```
```suggestion
                  <li>If |record|.{{[[ActiveLocks]]}} [=list/is empty=], then
```

rakuco replied:
Fixed, thanks.

domenic reviewed (commented):
Looking very solid; just some minor issues with nesting.

marcoscaceres reviewed (approved):
LGTM.

marcoscaceres commented:
@rakuco, any blockers or anything you need help with here?

rakuco commented:
> @rakuco, any blockers or anything you need help with here?

Nope, thanks a lot for reviewing and accepting the PR, and sorry I haven't merged it yet. I've been dragged into other stuff but will try to check if any tests need to be adjusted and merge it today or tomorrow.

marcoscaceres commented:
Great to hear! thanks for the update, @rakuco!

rakuco commented:
I'd like to thank everyone once again for reviewing this patch and for all the discussions. I've updated the PR message and included links to the Chromium bug and WPT PR that this change has originated. Since there have been no objections and the PR in WPT is close to being merged, I'm finally going to merge this one.
